src/runtime/rust/src/runtime/array.rs

```diff
@@ -0,0 +1,393 @@
+use std::{
+  any::TypeId,
+  convert::TryFrom,
+  mem,
+  os::raw::{c_int, c_void},
+  ptr,
+  slice,
+};
+
+use ndarray;
+
+use super::allocator::Allocation;
+use errors::*;
+use ffi::runtime::{
+  DLContext, DLDataType, DLDataTypeCode_kDLFloat, DLDataTypeCode_kDLInt, DLDataTypeCode_kDLUInt,
+  DLDeviceType_kDLCPU, DLTensor,
+};
+
+#[derive(PartialEq)]
+pub enum Storage<'a> {
+  Owned(Allocation),
+  View(&'a mut [u8], usize), // ptr, align
+}
+
+impl<'a> Storage<'a> {
+  pub fn new(size: usize, align: Option<usize>) -> Result<Storage<'static>> {
+    Ok(Storage::Owned(Allocation::new(size, align)?))
+  }
+
+  pub fn as_mut_ptr(&self) -> *mut u8 {
+    match self {
+      Storage::Owned(alloc) => alloc.as_mut_ptr(),
+      Storage::View(slice, _) => slice.as_ptr() as *mut u8,
+    }
+  }
+
+  pub fn size(&self) -> usize {
+    match self {
+      Storage::Owned(alloc) => alloc.size(),
+      Storage::View(slice, _) => slice.len(),
+    }
+  }
+
+  pub fn align(&self) -> usize {
+    match self {
+      Storage::Owned(alloc) => alloc.align(),
+      Storage::View(_, align) => *align,
+    }
+  }
+
+  pub fn as_ptr(&self) -> *const u8 {
+    self.as_mut_ptr() as *const _
+  }
+
+  pub fn view(&self) -> Storage<'a> {
+    match self {
+      Storage::Owned(alloc) => Storage::View(
+        unsafe { slice::from_raw_parts_mut(alloc.as_mut_ptr(), self.size()) },
+        self.align(),
+      ),
+      Storage::View(slice, _) => Storage::View(
+        unsafe { slice::from_raw_parts_mut(self.as_mut_ptr(), slice.len()) },
+        self.align(),
+      ),
+    }
+  }
+
+  pub fn is_owned(&self) -> bool {
+    match self {
+      Storage::Owned(_) => true,
+      _ => false,
+    }
+  }
+
+  pub fn to_owned(&self) -> Storage<'static> {
+    let s = Storage::new(self.size(), Some(self.align())).unwrap();
+    unsafe {
+      s.as_mut_ptr()
+        .copy_from_nonoverlapping(self.as_ptr(), self.size())
+    }
+    s
+  }
+}
+
+impl<'a, T> From<&'a [T]> for Storage<'a> {
+  fn from(data: &'a [T]) -> Self {
+    let data = unsafe {
+      slice::from_raw_parts_mut(
+        data.as_ptr() as *const u8 as *mut u8,
+        data.len() * mem::size_of::<T>() as usize,
+      )
+    };
+    Storage::View(data, mem::align_of::<T>())
+  }
+}
+
+#[derive(PartialEq)]
+pub struct Tensor<'a> {
+  pub(super) data: Storage<'a>,
+  pub(super) ctx: TVMContext,
+  pub(super) dtype: DataType,
+  pub(super) shape: Vec<usize>,
+  pub(super) strides: Option<Vec<usize>>,
+  pub(super) byte_offset: isize,
+  pub(super) numel: usize,
+  pub(super) dshape: Vec<i64>,
```

Comment:
this is the `dshape` mentioned in the PR description. When the target arch is 32-bit, `usize` is 32 bits. SInce the TVM module wants a 64-bit shape array, there needs to be an owned copy stored somewhere (i.e. in `dshape`).